server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/DefaultMetaRepository.java

```diff
@@ -125,57 +152,124 @@ public CompletableFuture<Mirror> mirror(String id, Revision revision) {
                 }
                 return mirror;
             });
         });
     }
 
+    private EntryNotFoundException mirrorNotFound(Revision revision, String mirrorFile) {
+        return new EntryNotFoundException(
+                "failed to find a mirror config for '" + mirrorFile + "' in " +
+                parent().name() + '/' + name() + " (revision: " + revision + ')');
+    }
+
     private CompletableFuture<List<Mirror>> allMirrors() {
-        return find(PATH_MIRRORS + "*.json").thenCompose(entries -> {
-            if (entries.isEmpty()) {
-                return UnmodifiableFuture.completedFuture(ImmutableList.of());
+        return find(NEW_MIRRORS_PATH + "*.json").thenCompose(this::handleAllMirrors);
+    }
+
+    private CompletableFuture<List<Mirror>> allMirrors(String repoName) {
+        return find("/repos/" + repoName + "/mirrors/*.json").thenCompose(this::handleAllMirrors);
+    }
+
+    private CompletableFuture<List<Mirror>> handleAllMirrors(Map<String, Entry<?>> entries) {
+        if (entries.isEmpty()) {
+            return UnmodifiableFuture.completedFuture(ImmutableList.of());
+        }
+
+        final Set<String> repoNames = new HashSet<>();
+        boolean hasProjectCredential = false;
+        final List<MirrorConfig> mirrorConfigs = toMirrorConfigs(entries);
+        for (MirrorConfig mirrorConfig : mirrorConfigs) {
+            if (mirrorConfig.repoCredential()) {
+                repoNames.add(mirrorConfig.localRepo());
+            } else {
+                hasProjectCredential = true;
             }
+        }
 
-            return projectCredentials().thenApply(credentials -> {
-                try {
-                    return parseMirrors(entries, credentials);
-                } catch (JsonProcessingException e) {
-                    return Exceptions.throwUnsafely(e);
-                }
-            });
-        });
+        return convert(new ArrayList<>(repoNames), hasProjectCredential,
+                       (repoCredentials, projectCredentials) -> {
+                           return mirrorConfigs.stream()
+                                               .map(mirrorConfig -> mirrorConfig.toMirror(
+                                                       parent(), repoCredentials, projectCredentials))
+                                               .filter(Objects::nonNull)
+                                               .collect(toImmutableList());
+                       });
     }
 
-    private List<Mirror> parseMirrors(Map<String, Entry<?>> entries, List<Credential> credentials)
-            throws JsonProcessingException {
+    private <T> CompletableFuture<T> convert(
+            List<String> credentialRepoNames,
+            boolean hasProjectCredential,
+            BiFunction<Map<String, List<Credential>>, List<Credential>, T> convertFunction) {
+        final ImmutableList<CompletableFuture<List<Credential>>> repoCredentialFutures;
```

Comment:
Personally, I prefer the extra IO instead of all of this complexity. Having said this, I guess the current code is fine as well.